### ISSUE TITLE
Should the language be more strict about strings and floating points numbers?

### ISSUE THREAD

**Issue by Dich0tomy** [hjson/hjson#129]
# Context
Hello.
I'm currently writing an application that uses Hjson as a manifest format.
I have a slight problem with it, though.

# Problem
My application specifically uses hjson as a format for a manifest which maps resources to versions, something like:
```
foo: 0.1.0
```

Now, a valid version doesn't necessarily condone to semver, it can also be `0.1` and such.

The problem is that this becomes sort of a semantic-parsing ambiguity.
```
foo: 0.1.0
```
will get parsed properly as
```json
{
  "foo": "0.1.0"
}
```
but remove the fix and we have:
```json
{
  "foo": 0.1
}
```

Even the hjson playground highliter tries to recognize `0.1.0` as a number, which probably also means there's some kind of disparity between the formal grammar and the actual generator (or the playground just doesn't condone to the formal grammar at all and regexes all of it?)
![image](https://github.com/hjson/hjson/assets/75364377/be73c4f7-3e0d-4f2f-92e6-958ad8bfecda)

# Proposed solution
I think the best way to resolve this is by making constructs like `\d\.\d(\.\d)+` ambiguous and forcing quoting for them.

The other solution would be forcing **everything** after `:` be quoted, that would get rid of all ambiguities of that kind, but I can imagine this can be too big of a breaking change and remove one of the main advertised features of this format.

I'm open for discussion.

**Comment by trobro** [comment]
Hi,

unfortunately the highlighter does not entirely follow the syntax specifications, PRs are welcome.

Hjson has been around for many years now so not breaking existing behavior is a top priority. We can therefore not introduce a requirement for constructs like `\d\.\d(\.\d)+` to be quoted, that could cause problems for many existing uses of Hjson.

How to best handle this problem depends on which implementation of Hjson you use. For example, hjson-go has an option `UseJSONNumber` which retains the exact string for any value decoded as a number.

**Comment by Dich0tomy** [comment]
> Hjson has been around for many years now so not breaking existing behavior is a top priority

Yeah I expected that answer. That's totally sensible and understandable.

> How to best handle this problem depends on which implementation of Hjson you use. For example, hjson-go has an option UseJSONNumber which retains the exact string for any value decoded as a number.

That's very nice to hear, but since this is implementation dependent and not reflected in the grammar, users will still see `0.1` highlighted differently than `0.1.0`, yeah?
